azure/scope/cluster.go

```diff
@@ -220,24 +221,33 @@ func (s *ClusterScope) RouteTableSpecs() []azure.RouteTableSpec {
 	return routetables
 }
 
-// NatGatewaySpecs returns the node nat gateway.
-func (s *ClusterScope) NatGatewaySpecs() []azure.NatGatewaySpec {
-	var natGateways []azure.NatGatewaySpec
+// NatGatewaySpecs returns the node NAT gateway.
+func (s *ClusterScope) NatGatewaySpecs() []azure.ResourceSpecGetter {
+	natGateways := make(map[string]azure.ResourceSpecGetter)
 
-	// We ignore the control plane nat gateway, as we will always use a LB to enable egress on the control plane.
+	// We ignore the control plane NAT gateway, as we will always use a LB to enable egress on the control plane.
 	for _, subnet := range s.NodeSubnets() {
 		if subnet.IsNatGatewayEnabled() {
-			natGateways = append(natGateways, azure.NatGatewaySpec{
-				Name: subnet.NatGateway.Name,
-				NatGatewayIP: infrav1.PublicIPSpec{
-					Name: subnet.NatGateway.NatGatewayIP.Name,
-				},
-				Subnet: subnet,
-			})
+			if _, ok := natGateways[subnet.NatGateway.Name]; !ok {
+				natGateways[subnet.NatGateway.Name] = &natgateways.NatGatewaySpec{
+					Name:           subnet.NatGateway.Name,
+					ResourceGroup:  s.ResourceGroup(),
+					SubscriptionID: s.SubscriptionID(),
+					Location:       s.Location(),
+					NatGatewayIP: infrav1.PublicIPSpec{
+						Name: subnet.NatGateway.NatGatewayIP.Name,
+					},
+				}
+			}
 		}
 	}
 
-	return natGateways
+	result := make([]azure.ResourceSpecGetter, 0, len(natGateways))
```

Comment:
suggestion for a small optimization in time and space: instead of having to loop through twice and storing the list of azure.ResourceSpecGetter twice (once in `natGateways` and once in `result`), you could instead define a `natGatewaysSet := map[string]struct{}` (cheaper since you're not storing the objects in there) and directly store the NatGatewaySpecs in the return slice. Each time you add a NAT GW to the specs you also add it to your set. And you use the set to determine if you need to add a nat GW to the specs for each subnet. That way you only loop once through the subnets.

```
for _, subnet := range s.NodeSubnets() {
	if subnet.IsNatGatewayEnabled() {
		if _, ok := natGatewaysSet[subnet.NatGateway.Name]; !ok {
			specs = append(specs, &natgateways.NatGatewaySpec{...})
			natGatewaysSet[subnet.NatGateway.Name] = struct{}{}
                }
        }
```